src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -397,12 +377,17 @@ private DocumentState SetParseOptions(ParseOptions options, bool onlyPreprocesso
 
     public DocumentState UpdateSourceCodeKind(SourceCodeKind kind)
     {
-        if (this.ParseOptions == null || kind == this.SourceCodeKind)
+        if (kind == SourceCodeKind)
         {
             return this;
         }
 
-        return this.SetParseOptions(this.ParseOptions.WithKind(kind), onlyPreprocessorDirectiveChange: false);
+        if (ParseOptions != null)
+        {
+            return UpdateParseOptionsAndSourceCodeKind(ParseOptions.WithKind(kind), onlyPreprocessorDirectiveChange: false);
+        }
+
+        return UpdateAttributes(Attributes.With(sourceCodeKind: kind));
```

Comment:
>return UpdateAttributes(Attributes.With(sourceCodeKind: kind)); [](http://example.com/codeflow?start=8&length=63)

This doesn't need to happen if the ParseOptions are set?

Reply:
Oh, I see. UpdateParseOptionsAndSourceCodeKind does it when it creates the new DocumentState.